docker/config/patch.diff

```diff
@@ -155,53 +312,256 @@ index 0000000..029493e
 +        pass
 +
 +    def initialize_webapp(self, config, webapp):
-+        pattern = r'/user/[^/]+/geotrellis/([0-9]+)/([^/]+)/([0-9]+)/([0-9]+)/([0-9]+)\.png.*'
-+        webapp.add_handlers(r'.*', [(pattern, GeoTrellisHandler)])
++        pattern = r'/user/[^/]+/geotrellis/([0-9]+)/([0-9]+)/([0-9]+)/([0-9]+)\.png.*'
++        webapp.add_handlers(r'.*', [(pattern, GeoTrellisTileHandler)])
++        pattern = r'/user/[^/]+/geotrellis/([0-9]+)/shutdown*'
++        webapp.add_handlers(r'.*', [(pattern, GeoTrellisShutdownHandler)])
 +
 +    def get_params(self, name, data, **kwargs):
 +        return {}
 +
-+    def disgorge(self, name):
-+        url = "http://localhost:%s/remove/%s" % (self.port, name)
-+        response = requests.get(url)
-+        status_code = response.status_code
-+        print(status_code)
-+        return status_code
++    def disgorge(self, name, **kwargs):
++        inproc_server_states = kwargs.pop('inproc_server_states', None)
++        if inproc_server_states is None:
++            raise Exception(
++                "GeoTrellis vis server requires kernel_id as kwarg to disgorge!")
++        if 'geotrellis' in inproc_server_states:
++            if name in inproc_server_states['geotrellis']['ports']:
++                port = inproc_server_states['geotrellis']['ports'][name]
++                url = "http://localhost:8000/user/hadoop/geotrellis/%s/shutdown" % port
++                response = requests.get(url)
++                status_code = response.status_code
++                inproc_server_states['geotrellis']['ports'].pop(name, None)
++            return None
++        return None
 +
 +    def ingest(self, data, name, **kwargs):
 +        from geopyspark.geotrellis.rdd import RasterRDD, TiledRasterRDD
 +        from geopyspark.geotrellis.render import PngRDD
++        from geopyspark.geotrellis.constants import ZOOM
++
++        inproc_server_states = kwargs.pop('inproc_server_states', None)
++        if inproc_server_states is None:
++            raise Exception(
++                "GeoTrellis vis server requires kernel_id as kwarg to ingest!")
++
++        if not "geotrellis" in inproc_server_states:
++            inproc_server_states["geotrellis"] = { "ports" : {} }
++
++        server_port = randint(49152, 65535)
++        while server_port in inproc_server_states['geotrellis']['ports'].values():
```

Comment:
Does this work when there is more than one document open (and/or more than one user)?